pkg/gameserverallocations/allocation_cache_test.go

```diff
@@ -0,0 +1,343 @@
+		Status: agonesv1.GameServerStatus{NodeName: "node1", State: agonesv1.GameServerStateReady, Players: &agonesv1.PlayerStatus{
+			Count:    2,
+			Capacity: 10,
+		}},
+	}
+
+	fixtures := map[string]struct {
+		list     []agonesv1.GameServer
+		test     func(*testing.T, []*agonesv1.GameServer)
+		features string
+	}{
+		"most allocated": {
+			// node1: 1 ready, 1 allocated, node2: 1 ready
+			list: []agonesv1.GameServer{gs1, gs2, gs3},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 2)
+				if !assert.Equal(t, []*agonesv1.GameServer{&gs1, &gs2}, list) {
+					for _, gs := range list {
+						logrus.WithField("name", gs.Name).Info("game server")
+					}
+				}
+			},
+			features: fmt.Sprintf("%s=false", runtime.FeatureStateAllocationFilter),
+		},
+		"allocated first (StateAllocationFilter)": {
+			list:     []agonesv1.GameServer{gs1, gs2, gs3},
+			features: fmt.Sprintf("%s=true", runtime.FeatureStateAllocationFilter),
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 3)
+				assert.Equal(t, "gs3", list[0].ObjectMeta.Name)
+				assert.Equal(t, "gs1", list[1].ObjectMeta.Name)
+				assert.Equal(t, "gs2", list[2].ObjectMeta.Name)
+			},
+		},
+		"nil player status (PlayerAllocationFilter)": {
+			list: []agonesv1.GameServer{gs1, gs2, gs4},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 3)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[0].ObjectMeta.Name)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[1].ObjectMeta.Name)
+				assert.Equal(t, &gs1, list[2])
+			},
+		},
+		"least player capacity (PlayerAllocationFilter)": {
+			list:     []agonesv1.GameServer{gs5, gs6},
+			features: fmt.Sprintf("%s=true", runtime.FeaturePlayerAllocationFilter),
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				require.Len(t, list, 2)
+				assert.Equal(t, "gs6", list[0].ObjectMeta.Name)
+				assert.Equal(t, "gs5", list[1].ObjectMeta.Name)
+			},
+		},
+		"list ready": {
+			// node1: 1 ready, node2: 2 ready
+			list: []agonesv1.GameServer{gs1, gs2, gs4},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 3)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[0].ObjectMeta.Name)
+				assert.Contains(t, []string{"gs2", "gs4"}, list[1].ObjectMeta.Name)
+				assert.Equal(t, &gs1, list[2])
+			},
+		},
+		"lexicographical (node name)": {
+			list: []agonesv1.GameServer{gs2, gs1},
+			test: func(t *testing.T, list []*agonesv1.GameServer) {
+				assert.Len(t, list, 2)
+				if !assert.Equal(t, []*agonesv1.GameServer{&gs1, &gs2}, list) {
+					for _, gs := range list {
+						logrus.WithField("name", gs.Name).Info("game server")
+					}
+				}
+			},
+		},
+	}
+
+	for k, v := range fixtures {
+		k := k
+		v := v
+		t.Run(k, func(t *testing.T) {
+			// deliberately not resetting the Feature state, to catch any possible unknown regressions with the
```

Comment:
Yes (admittedly this is a movement of an old test, as is much of this file) - since we don't know what features will be enabled/disabled across the whole set of features, we rely on go's random ordering of tests to find issues with combinations of feature flags over time.

Testing for every combination of feature flags everywhere for past and future flags is a combinatorial explosion, and probably not worth the effort (unless you have a clever way of doing this that I can't think of - in which case I'm all ears!)

Reply:
I don't think we need to test every combination everywhere, but it places where we have two explicit code paths and it's easy to add coverage (by adding a new table driven test entry) it seems like not too much work to ensure that all test runs do both old and new.